### DISCUSSION TITLE
NextJS import global stylesheets in all components

### DISCUSSION THREAD

**Discussion by Norfeldt** [cypress-io/cypress#16294] [Component Testing] [answered]
Thanks to @lmiller1990 I got c[ypress component testing working with my NextJS application](https://github.com/cypress-io/cypress/discussions/16136). It also plays nicely with `.scss` files. 

I just ran into an issue with my global styles which are imported in the `/pages/_app.tsx` - which is the only place NextJS wanna see them. Then you have to use css modules in all of your components. Really fine concept 👍 but my component tests should also have the global styles.

Tried to add them like this

> ./cypress/support/index.ts

```ts
/// <reference types="cypress" />

import '@testing-library/cypress/add-commands'

import '~/styles/normalize.scss'
import '~/styles/globals.scss'
import '~/styles/typography.scss'
```

But that throws off NextJS 

```sh
ERROR in ./styles/globals.scss
Module Error (from ./node_modules/next/dist/build/webpack/loaders/error-loader.js):
Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js. Or convert the import to Component-Level CSS (CSS Modules).
Read more: https://nextjs.org/docs/messages/css-global
...
```

**Comment by Norfeldt** [comment]
I found a "hack" for it.. Converted my global stylesheets to styled components 

> ./pages/_app.tsx

```tsx
import NormalizeStyle from '~/styles/NormalizeStyle'
import GlobalStyle from '~/styles/GlobalStyle'
import TypographyStyle from '~/styles/TypographyStyle'

function MyApp({ Component, pageProps }) {
  return (
    <>
      <NormalizeStyle />
      <GlobalStyle />
      <TypographyStyle />
   
      <Component {...pageProps} />
    </>
  )
}

export default MyApp
```

that allows me to get the styles in like this

```tsx
describe('DishCard', () => {
  it('looks great', () => {
    mount(
      <>
        <NormalizeStyle />
        <GlobalStyle />
        <TypographyStyle />

        <CartProvider>
          <MenuProvider>
            <DishCard id="dish02" />
          </MenuProvider>
        </CartProvider>
      </>
    )
    cy.findByText('CHICKEN', { exact: false })
  })
})
```

Currently trying to make a helper method to hide it a bit

> ./cypress/support/render.tsx

```tsx
import React from 'react'
import { mount } from '@cypress/react'

import NormalizeStyle from '~/styles/NormalizeStyle'
import GlobalStyle from '~/styles/GlobalStyle'
import TypographyStyle from '~/styles/TypographyStyle'

function render(Component: React.ReactNode): React.ReactNode {
  return mount(
    <>
      <NormalizeStyle />
      <GlobalStyle />
      <TypographyStyle />

      <Component />
    </>
  )
}

export { render }
```

```tsx
describe('DishCard', () => {
  it('looks great', () => {
    render(
      <>
        <CartProvider>
          <MenuProvider>
            <DishCard id="dish02" />
          </MenuProvider>
        </CartProvider>
      </>
    )
    cy.findByText('CHICKEN', { exact: false })
  })
})
```

but it does not work

```
Error
The following error originated from your test code, not from Cypress.

  > Element type is invalid: expected a string (for built-in components) or a class/function (for composite components) but got: object.

When Cypress detects uncaught errors originating from your test code it will automatically fail the current test.
```

**Reply by alfupe** [reply]
I would love Cypress to add the solution you've proposed here: https://github.com/cypress-io/cypress/discussions/16294#discussioncomment-708919

In the meantime, I've done a very silly mount alias that imports my global styles which import other things like semantic-ui, etc. (this is being used in a CRA application).

```js
import { mount as cyMount } from '@cypress/react';
import 'path/to/index.scss';

export function mount(Component, options) {
  return cyMount(Component, options);
}
```

and it's used like the cypress mount

```js
import PasswordRevealerInput from './PasswordRevealerInput';
import { mount } from 'modules/mount';

it('renders learn react link', () => {
  mount(<PasswordRevealerInput />);
  cy.get('input').type('Lorem ipsum');
  cy.get('input').should('have.attr', 'type').and('eq', 'password');
  cy.get('.eye').click();
  cy.get('input').should('have.attr', 'type').and('eq', 'text');
});
```

**Reply by Norfeldt** [reply]
I did try something similar without luck. I just retried your suggestion and it stills fails.

```ts
import { mount as cyMount } from '@cypress/react'
import '~/styles/Global.scss'

export function mount(Compontent: Parameters<typeof cyMount>[0], options?: Parameters<typeof cyMount>[1]) {
  return cyMount(Compontent, options)
}
```

<details>
<summary>Terminal error message</summary>

```sh
Module Error (from ./node_modules/next/dist/build/webpack/loaders/error-loader.js):
Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js. Or convert the import to Component-Level CSS (CSS Modules).
Read more: https://nextjs.org/docs/messages/css-global
Location: cypress/helpers/mountWithGlobalStyles/index.ts
Error: Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js. Or convert the import to Component-Level CSS (CSS Modules).
Read more: https://nextjs.org/docs/messages/css-global
Location: cypress/helpers/mountWithGlobalStyles/index.ts
    at Object.ErrorLoader (/Users/norfeldt/Repos/locali-web/node_modules/next/build/webpack/loaders/error-loader.ts:20:15)
 @ ./cypress/helpers/mountWithGlobalStyles/index.ts 2:0-30
 @ ./components/PaymentSucceededMessage/index.spec.tsx 9:0-64 16:4-9
 @ ./node_modules/@cypress/webpack-dev-server/dist/loader.js!./node_modules/@cypress/webpack-dev-server/dist/browser.js 37:16-141
 @ ./node_modules/@cypress/webpack-dev-server/dist/browser.js 3:4-41

ERROR in ./styles/Global.scss
Module build failed: Error: Final loader (./node_modules/next/dist/build/webpack/loaders/error-loader.js) didn't return a Buffer or String
    at processResult (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:60219:17)
    at /Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:60305:5
    at /Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25961:3
    at iterateNormalLoaders (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25787:10)
    at iterateNormalLoaders (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25794:10)
    at /Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25809:3
    at runSyncOrAsync (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25692:12)
    at iterateNormalLoaders (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25805:2)
    at Array.<anonymous> (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:25778:4)
    at runCallbacks (/Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:15395:15)
    at /Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:15568:4
    at /Users/norfeldt/Repos/locali-web/node_modules/next/dist/compiled/webpack/bundle5.js:24755:16
    at /Users/norfeldt/Library/Caches/Cypress/7.4.0/Cypress.app/Contents/Resources/app/packages/server/node_modules/graceful-fs/graceful-fs.js:90:16
    at FSReqCallback.readFileAfterClose [as oncomplete] (internal/fs/read_file_context.js:63:3)
 @ ./cypress/helpers/mountWithGlobalStyles/index.ts 2:0-30
 @ ./components/PaymentSucceededMessage/index.spec.tsx 9:0-64 16:4-9
```

</details>

**Reply by Norfeldt** [reply]
> Since Next.js 9.5.4, importing a CSS file from **node_modules** is permitted anywhere in your application.

I have found a "hack" - I can copy my global styles into `node_modules/@cypress/global.scss` and then this will work

```ts
import { mount as cyMount } from '@cypress/react'
import '@cypress/global.scss'

export function mount(
  tsx: Parameters<typeof cyMount>[0],
  options?: Parameters<typeof cyMount>[1]
) {
  return cyMount(tsx, options)
}
```

```tsx
import React from 'react'
// import { mount } from '@cypress/react'
import { mount } from '~/cypress/helpers/mountWithGlobalStyles'

import ResizeWrapper from '~/cypress/helpers/ResizeWrapper'
// import GlobalStyle from '~/styles/GlobalStyle'

import { CartProvider } from '~/state/CartContext'

import PaymentSucceededMessage from '.'

describe('PaymentYes', () => {
  it('displays', () => {
    mount(
      <>
        {/* <GlobalStyle /> */}
        <ResizeWrapper>
          <CartProvider>
            <PaymentSucceededMessage />
          </CartProvider>
        </ResizeWrapper>
      </>
    )

    cy.findByText('ORDER­BEKRÄFTELSE').should('be.visible')
  })
})
```

this can be automated by an script that does this copies it into the node_modules folder before starting the cypress component-testing.

**Reply by lmiller1990** [reply]
Interesting. I thought about this a bit more @alfupe @Norfeldt.

Something like

```json
{
  "component: {
    globalStylesheets: [/*...*/]
  }
}
```

Seems attractive, but I think it's a hard sell to the rest of the team. It leaks component testing configuration into Cypress at large. `cypress.json` is more for general, Cypress configuration. I think what might be better is configuring this on the React level, eg:

```js
import { mount, configure } from '@cypres/react'

configure({
  globalStylesheets: [/* ... */]
})
```

This might go in `cypress/support/index.js`, for example. Same ideal, but the coupling is with React, not the entire Cypress config (which is really huge and complex and goes through a lot of pre-processing/transformations). 

As for opting out, that's another issue, but should be solvable on the React level, too.

On the other hand, GitHub discussions are very difficult to follow. The UI is really confusing! I copy pasted this idea [here](https://github.com/cypress-io/cypress/issues/16509#issuecomment-852626005) so it won't get lost.

**Reply by Norfeldt** [reply]
> On the other hand, GitHub discussions are very difficult to follow. The UI is really confusing!

I totally agree with you @lmiller1990 !

**Comment by lmiller1990** [comment]
Hmm good point. This is going to be a problem for a lot of frameworks (not just Next.js). Importing them to support should generally be okay, but seems Next.js isn't happy.

I can look next week sometime, we should find a clean solution and build it into the nextjs adapter so people don't need to think about this.

**Reply by Norfeldt** [reply]
since this post I saw that @bahmutov demonstrate how to apply global styles https://glebbahmutov.com/blog/my-vision-for-component-tests/#components-all-the-way-down

Personally I would prefer to explicit exclude global styles and implicit have them..

**Reply by lmiller1990** [reply]
What do you mean by explicit exclude global styles? What kind of API do you have in mind?

I think we should have some way to modify the default `index.html` and include global styles (like `preview-head.html` in storybook). What do you think about this?

**Reply by Norfeldt** [reply]
I follow your idea about using a familiar convention done by story book 

https://storybook.js.org/docs/react/configure/story-rendering

Then we are just back to having to turn it on on every test (explicit) rather than having to turn it off for the instances where its not needed (implicit)

I would imagine an API something ala:

> cypress.json

```json
{
  "testFiles": "!(node_modules)/**/*.spec.{js,ts,jsx,tsx}",
  "componentFolder": "./",
  "globalStyleSheets": ["styles/global.scss"]
}
```

or if you need more fine grain control:

> cypress.json

```json
{
  "testFiles": "!(node_modules)/**/*.spec.{js,ts,jsx,tsx}",
  "componentFolder": "./",
  "globalStyleSheets": ["styles/general.scss", "normalize.scss", "typography.scss"]
}
```

Then they will be applied when doing 

```jsx
mount(<ComponentToTest/>)
```

and when global stylesheets should not be applied (could be a case for making reusable components across projects)

```jsx
mount(<ProjectIndependentComponentToTest />, {globalStyleSheets: false})
```

What do you think about that such an approach @lmiller1990 ?

**Reply by Norfeldt** [reply]
BTW the `stylesheets` option on mount does not work for me like it does from @bahmutov in his fine demonstration. Don't know if it has anything to do with using `.scss` 🤷‍♂️

**Reply by lmiller1990** [reply]
Looking at the source for stylesheets, it just puts them into a `<link>`: https://github.com/cypress-io/cypress/blob/1e7a3faa4c26bea5aea12be4f5eb023235c3b1a4/npm/mount-utils/src/index.ts#L78

So that is probably why it's not working with scss. You should be able to do an `import './foo.scss'` at the top of your test, though.

I think we should support the same API as storybook - expose a `preview-head.html` where you can put your global styles. The use case for a shared component across many projects is interesting, but in that case, you'd just have a different `preview-head.html` per project.

Some work is happening around supporting Storybook like APIs in Cypress, see here: https://github.com/cypress-io/cypress/issues/16379

**Reply by Norfeldt** [reply]
> So that is probably why it's not working with scss. You should be able to do an import './foo.scss' at the top of your test, though.

Can't do that because NextJS has set their webpack/config so it throws an error if you try to import a stylesheet that is not a css module. The only acceptable place to do this type of global style import is in the `./pages/_app.tsx`


> I think we should support the same API as storybook - expose a preview-head.html where you can put your global styles. The use case for a shared component across many projects is interesting, but in that case, you'd just have a different preview-head.html per project.

I trust you know the best solution since I believe you have the overview and know the implications. 

> Some work is happening around supporting Storybook like APIs in Cypress, see here: #16379

This link is a bit recursive 😂

**Reply by lmiller1990** [reply]
> Can't do that because NextJS has set their webpack/config so it throws an error if you try to import a stylesheet that is not a css module. The only acceptable place to do this type of global style import is in the ./pages/_app.tsx

Ah, I see. Next.js is painfully opinionated. Okay, I'll think about how we can solve this...

**Reply by Norfeldt** [reply]
For now I have to use styled components as a hack - would prefer to avoid it since I think that `.scss` and css modules does the trick for me.

**Reply by lmiller1990** [reply]
Agreed styled-components are a bit of a hack. I can't look at this until at least next week, most likely, if you have any other ideas in the meantime how we might solve this, please let me know. 

We might need to patch the next.js webpack config dynamically. CRA also has a similar problem where the webpack config has a bunch of limitations that make it painful to integrate, so it would be cool if we can find some kind of pattern that works for the general problem of "this wepback config is designed specifically for one thing, but we want to use it for something else".

**Reply by Norfeldt** [reply]
I have struggled with a lot of different ways to make the hack less painful. 

Was trying to keep my `.scss` files and then just `@import` them into my styled component - it did not work.

**Reply by lmiller1990** [reply]
There are many such problems with webpack configs outside of the env they are designed for. I will track them here: https://github.com/cypress-io/cypress/issues/16509

We need some kind of general solution for this.

**Reply by Norfeldt** [reply]
I'm all for general solutions :-)

**Comment by Vadorequest** [comment]
> Global CSS cannot be imported from files other than your Custom <App>. Please move all global CSS imports to pages/_app.js.

This issue is not supposed to happen on Next.js v10, there was an update that now allows to import .css files from any component. What version are you using? Upgrading might fix it (I import .css from any component and it works well)

**Reply by Norfeldt** [reply]
I'm using `"next": "10.1.3"` and `"sass": "^1.32.8"` to convert my `.scss` files

**Reply by Norfeldt** [reply]
@Vadorequest you think it's because I use `.scss`?

**Reply by Vadorequest** [reply]
https://nextjs.org/blog/next-9-2#built-in-css-support-for-global-stylesheets supported since 9.2, but you're not the first having issues with it. Maybe open an issue on the Next.js repo?

**Reply by Norfeldt** [reply]
I tried @Vadorequest  https://github.com/vercel/next.js/issues/25448 it was closed and moved to discussions

**Reply by Norfeldt** [reply]
> Since Next.js 9.5.4, importing a CSS file from **node_modules** is permitted anywhere in your application.

It's because it's not located in `node_modules`...

**Reply by Vadorequest** [reply]
Good catch, I hadn't notice it was limites to `node_modules`. I'm not sure if it's a bug then, but definitely bad luck for you. :/

**Reply by Norfeldt** [reply]
Not completely out of luck https://github.com/cypress-io/cypress/discussions/16294#discussioncomment-808601

**Comment by Norfeldt** [comment, marked as answer]
This **hack** works better for me than the styles components (which keeps showing a warning because `@import`)

> cypress/helpers/mountWithGlobalStyles.tsx

```tsx
import { mount as cyMount } from '@cypress/react'
import '@cypress/react/global.css'

export function mount(tsx: Parameters<typeof cyMount>[0], options?: Parameters<typeof cyMount>[1]) {
  return cyMount(tsx, options)
}
```

> styles/global.css

```css
@import './normalize.css';
@import './general.css';
@import './typography.css';

```

```sh
$ yarn add -D copyfiles
```

> package.json

```json
  "scripts": {
    ...
    "cy:open-ct": "yarn copyfiles -EV './styles/*.css' ./node_modules/@cypress/react && cypress open-ct",
    ...
```

**Reply by Norfeldt** [reply]
This will of course not have any hot reloading for changes in the global styles...

**Comment by ukind** [comment]
I think I found another solution to allowing global.css to be imported to `component.ts`, I tested it, and work for me [cypress next js](https://github.com/cypress-io/cypress-component-testing-apps/blob/b015bed3a71934557364ab7f8ced1bb8ec2bb305/react-next12-ts/next.config.js)

**Reply by EandrewJones** [reply]
Care to share?
